packages/txservice/src/provider.ts

```diff
@@ -159,24 +179,68 @@ export class ChainRpcProvider {
    * @param hash The hexadecimal hash string of the transaction.
    * @param confirmations Optional parameter to override the configured number of confirmations
    * required to validate the receipt.
-   * @param timeout Optional timeout parameter to override the configured parameter.
+   * @param timeout Optional timeout parameter in ms to override the configured parameter.
    *
    * @returns The ethers TransactionReceipt, if mined, otherwise null.
    */
-  public confirmTransaction(
-    transaction: Transaction,
+  public async confirmTransaction(
+    transaction: OnchainTransaction,
     confirmations: number = this.confirmationsRequired,
     timeout: number = this.confirmationTimeout,
-  ): ResultAsync<providers.TransactionReceipt | null, TransactionError> {
-    return this.resultWrapper<providers.TransactionReceipt>(
-      true,
-      () => {
-        // The only way to access the functionality internal to ethers for handling replacement tx.
-        // See issue: https://github.com/ethers-io/ethers.js/issues/1775
-        return Promise.race(transaction.responses.map((response) => (response as any).wait(confirmations, timeout)));
-      },
-      false,
-    );
+  ): Promise<providers.TransactionReceipt> {
+    const start = Date.now();
+    // Using a timed out variable calculated at the end of the loop - this way we can be sure at
+    // least one iteration is completed here.
+    let timedOut = false;
+    while (!timedOut) {
+      const errors: NxtpError[] = [];
+      // Populate a list of promises to retrieve every receipt for every hash.
+      const _receipts: Promise<providers.TransactionReceipt | null>[] = transaction.responses.map(async (response) => {
+        try {
+          return await this.getTransactionReceipt(response.hash);
+        } catch (error) {
+          errors.push(error);
+          return null;
+        }
+      });
+      // Wait until all the 'receipts' (or errors) have been pushed to the list.
+      const receipts = await Promise.all(_receipts);
```

Comment:
why not just filter out the `null` return values here rather than handling it below?

Reply:
genius